src/pushin.ts

```diff
@@ -61,6 +64,17 @@ export class PushIn {
     }
   }
 
+  /**
+   * Does all necessary cleanups by removing event listeners.
+   */
+  destroy(): void {
+    cancelAnimationFrame(this.lastAnimationFrameId);
+
+    while (this.cleanupFns.length) {
+      this.cleanupFns.pop()!();
```

Comment:
In TypeScript this is called _non-null assertion operator_.

The return value of `pop()` is `T | undefined`. E.g. if it's an empty array then `pop()` will return `[].pop() -> undefined`, otherwise, it will return the last value.

When we do `this.cleanupFns.pop()()` w/o _non-null assertion operator_ the TS compiler says `Cannot invoke an object which is possibly 'undefined'.`. So it tells us that we should do a check if the value is not `undefined`, e.g. like this:
```ts
const cleanupFn: VoidFunction | undefined = this.cleanupFns.pop();
if (cleanupFn) {
  cleanupFn();
}
```
The reason why we don't have that check is because `pop()` is wrapped into `while (this.cleanupFns.length)`, which basically ensures that `pop()` will be executed only on **non-empty array** and the return value will not equal `undefined`.

Other examples:
```ts
function assertDefined(value) {
  if (value == null) {
    throw new Error(`Value is not defined: ${value}`);
  }
}

let _value: string | null = null;

export function getValue(): string {
  assertDefined(_value);
  return _value;
}
```

In the above example the TS compiler will throw an error `Type 'string | null' is not assignable to type 'string'.`, because `_value` is type of `string | null`, but the `getValue` return type is `string`. It tells us _hey, the return type is `string`, but you're returning `string | null`, I cannot allow that_.

We know that `assertDefined` throws an error if `_value` is `null` so `return _value` will actually always return `string` and not `string | null`. To fix that we should use the operator:
```ts
export function getValue(): string {
  assertDefined(_value);
  return _value !;
}
```

We're telling the compiler _I'm sure that value will not be nullable here_.

Recent TS versions introduced other features for such cases:
```ts
function assertDefined<T>(value: T): asserts value is NonNullable<T> {
  if (value == null) {
    throw new Error(`Value is not defined: ${value}`);
  }
}

let _value: string | null = null;

export function getValue(): string {
  assertDefined(_value);
  return _value; // We don't have to use the `!` operator here
}
```

`T` is a generic type which means we can pass anything to `assertDefined`, e.g. `string | number | null | object`, anything. `NonNullable` is a type that excludes `null | undefined` from `T`.

Reply:
Oh wow, very interesting. That is a great explanation, thank you!